wen-restart/src/heaviest_fork_aggregate.rs

```diff
@@ -127,36 +133,34 @@ impl HeaviestForkAggregate {
             shred_version: received_heaviest_fork.shred_version as u32,
             wallclock: received_heaviest_fork.wallclock,
         };
-        if let Some(old_heaviest_fork) = self
-            .heaviest_forks
-            .insert(*from, received_heaviest_fork.clone())
-        {
-            if Self::should_replace(&old_heaviest_fork, &received_heaviest_fork) {
-                let entry = self
-                    .block_stake_map
-                    .get_mut(&(
-                        old_heaviest_fork.last_slot,
-                        old_heaviest_fork.last_slot_hash,
-                    ))
-                    .unwrap();
-                info!(
-                    "{:?} Replacing old heaviest fork from {:?} with {:?}",
-                    from, old_heaviest_fork, received_heaviest_fork
+        let mut entry_exists = false;
+        if let Some(old_heaviest_fork) = self.heaviest_forks.get(from) {
+            entry_exists = true;
+            if old_heaviest_fork.last_slot != received_heaviest_fork.last_slot
+                || old_heaviest_fork.last_slot_hash != received_heaviest_fork.last_slot_hash
+            {
+                return HeaviestForkAggregateResult::DifferentVersionExists(
+                    old_heaviest_fork.clone(),
+                    received_heaviest_fork,
                 );
-                *entry = entry.saturating_sub(sender_stake);
-            } else {
-                return None;
             }
+            if !Self::should_replace(old_heaviest_fork, &received_heaviest_fork) {
+                return HeaviestForkAggregateResult::AlreadyExists;
+            }
```

Comment:
Together these seems like they should be merged into one common validation function on `RestartHeaviestFork` called `is_valid_change` that returns a `Result<HeaviestForkAggregateResult>` based on whether it's OK, or an error

Reply:
Done.